File: exercise/week04/ts/src/construct/constructInventory.ts
import { Artefact } from "./artefact";

export class ConstructInventory {
    private artefacts: Artefact[];

    constructor(artefacts: Artefact[]) {
        this.artefacts = artefacts;
    }

    updateSimulation(): void {
        for (const artefact of this.artefacts) {
            if (
                artefact.name !== "Aged Signal" &&
                artefact.name !== "Backdoor Pass to TAFKAL80ETC Protocol"
            ) {
                if (artefact.integrity > 0 && artefact.name !== "Sulfuras Core Fragment") {
                    artefact.integrity--;
                }
            } else {
                if (artefact.integrity < 50) {
                    artefact.integrity++;

                    if (artefact.name === "Backdoor Pass to TAFKAL80ETC Protocol") {
                        if (artefact.timeToLive < 11 && artefact.integrity < 50) {
                            artefact.integrity++;
                        }
                        if (artefact.timeToLive < 6 && artefact.integrity < 50) {
                            artefact.integrity++;
                        }
                    }
                }
            }

            if (artefact.name !== "Sulfuras Core Fragment") {
                artefact.timeToLive--;
            }

            if (artefact.timeToLive < 0) {
                if (artefact.name !== "Aged Signal") {
                    if (artefact.name !== "Backdoor Pass to TAFKAL80ETC Protocol") {
                        if (artefact.integrity > 0 && artefact.name !== "Sulfuras Core Fragment") {
                            artefact.integrity--;
                        }
                    } else {
                        artefact.integrity = 0;
                    }
                } else {
                    if (artefact.integrity < 50) {
                        artefact.integrity++;
                    }
                }
            }
        }
    }

    getArtefacts(): Artefact[] {
        return this.artefacts;
    }
}

File: exercise/week04/java/src/main/java/construct/ConstructInventory.java
package construct;

class ConstructInventory {
    Artefact[] artefacts;

    public ConstructInventory(Artefact[] artefacts) {
        this.artefacts = artefacts;
    }

    public void updateSimulation() {
        for (int i = 0; i < artefacts.length; i++) {
            Artefact artefact = artefacts[i];

            if (!artefact.name.equals("Aged Signal")
                    && !artefact.name.equals("Backdoor Pass to TAFKAL80ETC Protocol")) {
                if (artefact.integrity > 0) {
                    if (!artefact.name.equals("Sulfuras Core Fragment")) {
                        artefact.integrity = artefact.integrity - 1;
                    }
                }
            } else {
                if (artefact.integrity < 50) {
                    artefact.integrity = artefact.integrity + 1;

                    if (artefact.name.equals("Backdoor Pass to TAFKAL80ETC Protocol")) {
                        if (artefact.timeToLive < 11) {
                            if (artefact.integrity < 50) {
                                artefact.integrity = artefact.integrity + 1;
                            }
                        }

                        if (artefact.timeToLive < 6) {
                            if (artefact.integrity < 50) {
                                artefact.integrity = artefact.integrity + 1;
                            }
                        }
                    }
                }
            }

            if (!artefact.name.equals("Sulfuras Core Fragment")) {
                artefact.timeToLive = artefact.timeToLive - 1;
            }

            if (artefact.timeToLive < 0) {
                if (!artefact.name.equals("Aged Signal")) {
                    if (!artefact.name.equals("Backdoor Pass to TAFKAL80ETC Protocol")) {
                        if (artefact.integrity > 0) {
                            if (!artefact.name.equals("Sulfuras Core Fragment")) {
                                artefact.integrity = artefact.integrity - 1;
                            }
                        }
                    } else {
                        artefact.integrity = artefact.integrity - artefact.integrity;
                    }
                } else {
                    if (artefact.integrity < 50) {
                        artefact.integrity = artefact.integrity + 1;
                    }
                }
            }
        }
    }
}
